### PR TITLE
Checks: Plumb channel and improve robustness of diff urls

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
See https://github.com/web-platform-tests/wpt/issues/14465

This avoids version strings in a couple of cases.

### CODE REVIEW

File: shared/run_diff.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ import (
 type DiffAPI interface {
 	GetRunsDiff(before, after TestRun, filter DiffFilterParam, paths mapset.Set) (RunDiff, error)
 	GetDiffURL(before, after TestRun, diffFilter *DiffFilterParam) *url.URL
-	GetMasterDiffURL(sha string, product ProductSpec) *url.URL
+	GetMasterDiffURL(sha string, product ProductSpec, diffFilter *DiffFilterParam) *url.URL
```
Comment (Hexcles) [inline shared/run_diff.go]: Why the change of function signature? Is there (or will there be) any call of `GetMasterDiffURL` that specifies non-nil diff filter param?

File: shared/run_diff.go [hunk 2]
```diff
@@ -42,10 +42,13 @@ func NewDiffAPI(ctx context.Context) DiffAPI {
...
 func (d diffAPIImpl) GetDiffURL(before, after TestRun, diffFilter *DiffFilterParam) *url.URL {
 	filter := TestRunFilter{}
-	filter.Products = ProductSpecs{
-		ProductSpec{ProductAtRevision: before.ProductAtRevision},
-		ProductSpec{ProductAtRevision: after.ProductAtRevision},
-	}
+	filter.Products = make(ProductSpecs, 2)
+	filter.Products[0].BrowserName = before.BrowserName
+	filter.Products[0].Revision = before.Revision
+	filter.Products[0].Labels = mapset.NewSet(before.Channel())
```
Comment (Hexcles) [inline shared/run_diff.go]: What about including all the labels?

Comment (lukebjerring) [reply shared/run_diff.go]: Sigh. Let's land https://github.com/web-platform-tests/wpt.fyi/pull/906 and then it doesn't matter at all.

File: shared/run_diff.go (hunk 1 above)
Comment (lukebjerring) [reply shared/run_diff.go]: Consistency.

File: shared/models.go [hunk 3]
```diff
@@ -119,6 +119,20 @@ func (r TestRun) IsExperimental() bool {
...
 }
 
+// Channel return the channel label, if any, for the given run.
+func (r TestRun) Channel() string {
+	for _, label := range r.Labels {
+		switch label {
+		case StableLabel,
+			ExperimentalLabel,
+			DevLabel,
+			BetaLabel:
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply shared/models.go]: Yeah.. ok . Fixed.
A) #538 needs to be concluded
B) the query builders are wrong for a different, similar reason.

File: api/checks/update.go [hunk 4]
```diff
@@ -172,8 +172,12 @@ func getDiffSummary(aeAPI shared.AppEngineAPI, diffAPI shared.DiffAPI, baseRun,
 	checkState := summaries.CheckState{
 		TestRun: &headRun,
 		Product: shared.ProductSpec{
-			ProductAtRevision: headRun.ProductAtRevision,
-			Labels:            labels,
+			// [browser]@[sha] is plenty specific, and avoids bad version strings.
```
Comment (Hexcles) [inline api/checks/update.go]: This is probably not specific enough in some cases.

Example: when `headRun` is a PR run, we would only pass `browser[channel]@revision` to `GetMasterDiffURL`, which will match to both `pr_base` and `pr_head` runs, and the users would always expect the `pr_head` run.

File: api/checks/update.go [hunk 5]
```diff
@@ -172,8 +172,12 @@ func getDiffSummary(aeAPI shared.AppEngineAPI, diffAPI shared.DiffAPI, baseRun,
 	checkState := summaries.CheckState{
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/checks/update.go]: Done.

File: api/receiver/create_run.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/deckarep/golang-set"
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/receiver/create_run.go]: Done.

File: shared/params.go [hunk 7]
```diff
@@ -113,13 +113,18 @@ func (p ProductSpecs) Strings() []string {
...
-	if p.Labels != nil && p.Labels.Cardinality() > 0 {
-		labels := make([]string, 0, p.Labels.Cardinality())
-		for l := range p.Labels.Iter() {
-			labels = append(labels, l.(string))
+	if p.Labels != nil {
+		p.Labels.Remove("") // Remove the empty label, if present.
+		if p.Labels.Cardinality() > 0 {
+			labels := make([]string, 0, p.Labels.Cardinality())
+			for l := range p.Labels.Iter() {
+				if l != "" {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply shared/params.go]: Done.

File: api/checks/update.go (hunk 4 above)
Comment (lukebjerring) [reply api/checks/update.go]: This is deliberately the case. This only needs to identify any _runs_ precisely enough for a given check; i.e. the CheckState's product is specific enough to identify chrome[experimental], and what the means as a check depends on what's been run (e.g. `master` vs an earlier `master`, `pr_base` vs `pr_head`, etc.)

As for below, let's rejig the API to take a run to diff against master, and then we can be less ambiguous :)

**Comment by lukebjerring** [comment]
Double-checking behaviour @ https://github.com/lukebjerring/wpt/pull/8/checks?sha=6d14f4b50f4e9121bd55504a6cc33b471569eb2d
